### PR TITLE
Dynamic tabs: use buttons rather than links

### LINKED ISSUE #32945: Navs and tabs should look the same with <button>
Elements like navs and tabs are mostly used with `<a>` in the docs. Anyhow modern JavaScript apps (e.g. using React or Angular) often do not use `<a>` elements to trigger internal ui events.

IMO It should be possible to use this elements CSS with buttons, too and they should look the same:

```
<ul class="nav">
  <li class="nav-item">
    <button class="nav-link active">Active</a>
  </li>
  <li class="nav-item">
    <button class="nav-link">Link</a>
  </li>
  <li class="nav-item">
    <button class="nav-link">Link</a>
  </li>
</ul>
```

Currently using components like `.nav` with buttons will totally break the layout.

This may affect the CSS only and must not apply to Bootstrap's Javascript.

Target version: 4.x / 5.x

### PR DESCRIPTION
**Pull request by patrickhlauke**
- change docs
- tweak styles to neutralise border and background

this is more semantically correct. note that currently, the forced `:focus` in reboot styles is making this look rather ugly - see separate PR https://github.com/twbs/bootstrap/pull/32631 to remove that cargo culted (?) style from reboot.

Preview: https://deploy-preview-32630--twbs-bootstrap.netlify.app/docs/5.0/components/navs-tabs/#javascript-behavior

### CODE REVIEW

**Comment by patrickhlauke** [comment]
set to draft for now while we wait for https://github.com/twbs/bootstrap/pull/32631 - once that is in place (removing the forced `:focus` style on buttons), this works cleanly.

only minor concerns:

- `.nav-link` used on buttons may not seem logical, but changing the class name now would be a breaking change
- the buttons have black, rather than link-blue, text

**Review by MartijnCuppens — APPROVED** [review]

**Review by mdo — COMMENTED** [review]
We should include some docs mentions here for why we use `<button>` elements here.

**Comment by patrickhlauke** [comment]
> We should include some docs mentions here for why we use `<button>` elements here.

added a short sentence to explain. also took opportunity to update the visual and unit tests

File: scss/_nav.scss [hunk 1]
```diff
@@ -80,6 +81,8 @@
 
 .nav-pills {
   .nav-link {
+    background: none;
```
Comment (XhmikosR) [inline scss/_nav.scss:84]: Should the shorthand be used here and the other CSS change?

Also, not sure about `border: 0` / `border: none`

**Comment by ffoodd** [comment]
You may use `isolation: isolate` on the focus state to ensure the focused item appears above the active one:

**without** `isolation` |  **with** `isolation
Capture d’écran de 2021-01-07 10-58-30 | 
Capture d’écran de 2021-01-07 10-58-17

Simple improvement, might be used elsewhere since [support is very well aligned with our browsers list](https://caniuse.com/mdn-css_properties_isolation).

File: scss/_nav.scss (hunk 1 above)
Comment (patrickhlauke) [reply scss/_nav.scss:84]: we do use the `background` shorthand in a few places (to set things, not to neutralise them, admittedly).

ditto for `border: 0`. we use that in various places. no `border: none`

**Comment by patrickhlauke** [comment]
> You may use `isolation: isolate` on the focus state to ensure the focused item appears above the active one

on which element/class are you adding it to? i just messed around in Firefox devtools, and seems adding it to `.nav-link` and to `.nav-item`, but no joy...

**Comment by ffoodd** [comment]
On `.nav-link`'s hover/focus state, since they're the one holding borders.

**Comment by patrickhlauke** [comment]
@ffoodd good call on the isolation thing. done.

**Comment by patrickhlauke** [comment]
@mdo good to merge?

**Comment by patrickhlauke** [comment]
As the carousel buttons stuff has been merged, would be keen to see this make its way into main as well. Any concerns on this?

**Review by mdo — APPROVED** [review]
LGTM. I'm fine including in Beta 2 or holding for Beta 3.

**Review by ffoodd — APPROVED** [review]

**Comment by XhmikosR** [comment]
1. Do we have a test with JS test `a` tags?
2. I think the wording needs some tweaking:

  >Dynamic tabbed interfaces, as described in the WAI ARIA Authoring Practices, require role="tablist", role="tab", role="tabpanel", and additional aria- attributes in order to convey their structure, functionality and current state to users of assistive technologies (such as screen readers). As a best practice, we recommend using <button> elements for the tabs, as these are controls that trigger a dynamic change, rather than links that navigate to a new page or location.

  I feel like It's not clear that if someone is using `button`s, the `role` isn't needed.

**Comment by patrickhlauke** [comment]
> > Dynamic tabbed interfaces, as described in the [WAI ARIA Authoring Practices](https://www.w3.org/TR/wai-aria-practices/#tabpanel), require `role="tablist"`, `role="tab"`, `role="tabpanel"`, and additional `aria-` attributes in order to convey their structure, functionality and current state to users of assistive technologies (such as screen readers). As a best practice, we recommend using `<button>` elements for the tabs, as these are controls that trigger a dynamic change, rather than links that navigate to a new page or location.
> 
> I feel like It's not clear that if someone is using `button`s, the `role` isn't needed.

`role="tab"` *is* needed even when using `<button>`

**Comment by patrickhlauke** [comment]
> LGTM. I'm fine including in Beta 2 or holding for Beta 3.

ah darn, sorry...forgot about the whole beta 2 / beta 3 aspect...got confused with the release of 4.6.0 recently thinking "so why is this still hanging there and not being merged?"

ok, i'm happy either way, whether it makes it into 2 or 3

**Comment by patrickhlauke** [comment]
> 1. Do we have a test with JS test `a` tags?

https://github.com/twbs/bootstrap/pull/32630/files#diff-e808aa1f3605a07d2a11a7dcb2f81c71f6bd2fd9257979ed460a3f3141d07bc8R48

unless you meant replicating every test to run both for `<button>` and `<a>` elements?

**Comment by XhmikosR** [comment]
> 
> 
> > 1. Do we have a test with JS test `a` tags?
> 
> https://github.com/twbs/bootstrap/pull/32630/files#diff-e808aa1f3605a07d2a11a7dcb2f81c71f6bd2fd9257979ed460a3f3141d07bc8R48
> 
> unless you meant replicating every test to run both for `<button>` and `<a>` elements?

Not all of them, just one test like you did with that other PR of yours.

> `role="tab"` _is_ needed even when using `<button>`

Sorry, missed that due to horizontal scrolling :)

**Comment by patrickhlauke** [comment]
> > > 1. Do we have a test with JS test `a` tags?
> > 
> > 
> > https://github.com/twbs/bootstrap/pull/32630/files#diff-e808aa1f3605a07d2a11a7dcb2f81c71f6bd2fd9257979ed460a3f3141d07bc8R48
> > unless you meant replicating every test to run both for `<button>` and `<a>` elements?
> 
> Not all of them, just one test like you did with that other PR of yours.

So we're good here, as I did it for this PR as well as per link above :)

**Comment by XhmikosR** [comment]
Cool. One last thing. I wouldn't remove any tests. Even if something is inappropriate, if it's still supported, it should be tested.

**Comment by patrickhlauke** [comment]
> Cool. One last thing. I wouldn't remove any tests. Even if something is inappropriate, if it's still supported, it should be tested.

assuming it's about the dynamic tabs using dropdown menus, as that pattern is completely wrong/inappropriate (from an accessibility and usability pov), not sure it makes sense to keep it in. in fact, tempted for v5 to strip out any code that handles this (as we already warned about/deprecated it in v4)

File: js/tests/unit/tab.spec.js [hunk 2]
```diff
@@ -277,55 +301,6 @@ describe('Tab', () => {
 
       firstTab.show()
     })
-
```
Comment (patrickhlauke) [inline js/tests/unit/tab.spec.js:280]: this test seems to test something that was never documented as a kosher technique/structure? it's invalid markup (having buttons inside of links) and should never be used, and i don't think we show this in any docs/examples (checked old v2 and v3 docs too, and couldn't see any example of this). feels safe to remove?

Comment (rohit2sharma95) [reply js/tests/unit/tab.spec.js:280]: This one?
**_should not fire hidden when hide is prevented_**

Comment (patrickhlauke) [reply js/tests/unit/tab.spec.js:280]: yeah but not using close buttons this way. maybe programmatically removed ones?

Comment (patrickhlauke) [reply js/tests/unit/tab.spec.js:280]: i've put the code for this and the dropdown tabs back for now...will need some more careful work here (to reformulate this one into a programmatic handling of tabs disappearing/being removed, and the other one a "breaking" change that should happen when we remove the actual logic in tab.js that handles this as well)

File: js/tests/visual/tab.html [hunk 3]
```diff
@@ -159,47 +159,14 @@ <h4>Tabs with fade (no initially active pane)</h4>
         </div>
       </div>
 
-      <h4>Tabs with nav (with fade)</h4>
```
Comment (patrickhlauke) [inline js/tests/visual/tab.html:162]: Removed this as it's a weirdly invalid example, markup-wise. Not sure what this was supposed to show/visually test for...

**Comment by patrickhlauke** [comment]
Put back the two tests that I had originally shelved. This should keep this PR focused and non-breaking

File: js/tests/visual/tab.html (hunk 3 above)
Comment (XhmikosR) [reply js/tests/visual/tab.html:162]: Maybe just nav without ul?

**Review by mdo — APPROVED** [review]

File: js/tests/visual/tab.html (hunk 3 above)
Comment (patrickhlauke) [reply js/tests/visual/tab.html:162]: right, fixed up (using the suggested pattern from the example using `<nav>` in https://getbootstrap.com/docs/5.0/components/navs-tabs/#javascript-behavior, and removing the dropdown) and reintroduced the visual test. this is now good to go from my end i'd say

**Comment by patrickhlauke** [comment]
:rocket:

**Comment by MickL** [comment]
Has this been published with beta2 6 days ago? For me buttons still dont look good inside `nav-pills`:

Bildschirmfoto 2021-02-16 um 13 56 01

Using the exact code from the docs:

```
<ul class="nav nav-pills">
  <li class="nav-item">
    <button class="nav-link active" aria-current="page">Active</button>
  </li>
  <li class="nav-item">
    <button class="nav-link">Link</button>
  </li>
  <li class="nav-item">
    <button class="nav-link">Link</button>
  </li>
  <li class="nav-item">
    <button class="nav-link disabled" tabindex="-1" aria-disabled="true">Disabled</button>
  </li>
</ul>
```

**Comment by patrickhlauke** [comment]
@MickL if you're using the exact same code as in the docs, but it's looking different from https://getbootstrap.com/docs/5.0/components/navs-tabs/#javascript-behavior, then it's likely some other modification/customisation you're doing that's affecting this?

**Comment by MickL** [comment]
@patrickhlauke obviously it is exactly the same BUT i changed `<a>` for `<button>`

**Comment by ffoodd** [comment]
@MickL Our current docs example (linked by Patrick) is working fine with `button`, this is why this PR got merged and your issue closed.

If you still see any issue related to this in beta 21, please open a new issue—and stop spamming an already merged PR that we cannot re-open.

**Pull request merged**